src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -738,12 +961,192 @@ export class SavedObjectsRepository {
+      refresh,
+      ignore: [404],
+      body: {
+        doc,
+      },
+    });
+
+    if (updateResponse.status === 404) {
+      // see "404s from missing index" above
+      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+    }
+
+    return {};
+  }
+
+  async removeNamespaces(
+    type: string,
+    id: string,
+    namespaces: string[],
+    options: SavedObjectsRemoveNamespacesOptions = {}
+  ): Promise<{}> {
+    if (!this._allowedTypes.includes(type)) {
+      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+    }
+
+    if (!this._registry.isMultiNamespace(type)) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(`${type} doesn't support namespaces`);
+    }
+
+    if (!namespaces || !Array.isArray(namespaces) || !namespaces.length) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(
+        'namespaces must be a non-empty array of strings'
+      );
+    }
+
+    const { refresh = DEFAULT_REFRESH_SETTING } = options;
+
+    const rawId = this._serializer.generateRawId(undefined, type, id);
+    const getResponse = await this._callCluster('get', {
+      id: rawId,
+      index: this.getIndexForType(type),
+      ignore: [404],
+    });
+
+    const getDocNotFound = getResponse.found === false;
+    const getIndexNotFound = getResponse.status === 404;
+    if (getDocNotFound || getIndexNotFound) {
+      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+    }
+
+    const existingNamespaces = getResponse._source.namespaces as string[] | undefined;
+    if (existingNamespaces) {
+      // if there are somehow no existing namespaces, don't treat this as a missing namespace --
+      // instead, allow the operation to proceed and delete this saved object
+      const missing = namespaces.filter(x => !existingNamespaces.includes(x));
+      if (missing.length) {
+        throw SavedObjectsErrorHelpers.createBadRequestError(
+          `${rawId} doesn't exist in the following namespace(s): ${missing.join(', ')}`
+        );
+      }
+    }
+    const remainingNamespaces = existingNamespaces
+      ? existingNamespaces.filter(x => !namespaces.includes(x))
+      : [];
+
+    if (remainingNamespaces.length) {
+      // if there is 1 or more namespace remaining, update the saved object
+      const time = this._getCurrentTime();
+
+      const doc = {
+        updated_at: time,
+        namespaces: remainingNamespaces,
+      };
+
+      const updateResponse = await this._writeToCluster('update', {
+        id: rawId,
+        index: this.getIndexForType(type),
+        refresh,
+        ignore: [404],
+        body: {
+          doc,
+        },
+      });
+
+      if (updateResponse.status === 404) {
+        // see "404s from missing index" above
+        throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+      }
+      return {};
+    } else {
+      // if there are no namespaces remaining, delete the saved object
```

Comment:
This operation (`removeNamespaces`) initiates exactly two cluster calls, either `get`+`update` or `get`+`delete`. It seemed prudent here to just use two separate calls for simplicity's sake.

The `deleteByNamespace` operation, on the other hand, has to account for multiple saved object types. If we did not use painless for that, it would result in three cluster calls -- `mget`, `bulk` update (for multi-namespace objects that exist in multiple spaces), and `deleteByQuery` (for single-namespace objects, and multi-namespace objects that exist only in this space).